File: frontend/src/types/order.ts
/**
 * Order Management Type Definitions
 * Aligned with backend Order entity and DTOs
 */

export type OrderStatus =
  | 'RECEIVED'
  | 'PREPARING'
  | 'OVEN'
  | 'BAKED'
  | 'READY'
  | 'DISPATCHED'
  | 'DELIVERED'
  | 'COMPLETED'
  | 'SERVED'
  | 'CANCELLED';

export type OrderType =
  | 'DINE_IN'
  | 'TAKEAWAY'
  | 'DELIVERY';

export type PaymentStatus =
  | 'PENDING'
  | 'PAID'
  | 'FAILED'
  | 'REFUNDED';

export type PaymentMethod =
  | 'CASH'
  | 'CARD'
  | 'UPI'
  | 'WALLET';

export type OrderPriority =
  | 'NORMAL'
  | 'URGENT';

export interface OrderItem {
  menuItemId: string;
  name: string;
  quantity: number;
  price: number;
  variant?: string;
  customizations?: string[];
}

export interface DeliveryAddress {
  street: string;
  city: string;
  pincode: string;
  latitude?: number;
  longitude?: number;
}

export interface Order {
  id: string;
  orderNumber: string;
  customerId?: string;
  customerName: string;
  customerPhone?: string;
  storeId: string;
  items: OrderItem[];
  subtotal: number;
  deliveryFee: number;
  tax: number;
  total: number;
  totalAmount?: number; // Alias for total (for backwards compatibility)
  status: OrderStatus;
  orderType: OrderType;
  paymentStatus: PaymentStatus;
  paymentMethod?: PaymentMethod;
  paymentTransactionId?: string;
  priority: OrderPriority;
  preparationTime?: number;
  estimatedDeliveryTime?: string;
  deliveryAddress?: DeliveryAddress;
  assignedDriverId?: string;
  driverId?: string; // Alias for assignedDriverId (for backwards compatibility)
  driverName?: string; // Driver name for display
  specialInstructions?: string;

  // Timestamps
  createdAt: string;
  updatedAt: string;
  completedAt?: string;
  cancelledAt?: string;
  cancellationReason?: string;

  // Kitchen workflow timestamps
  receivedAt?: string;
  preparingStartedAt?: string;
  ovenStartedAt?: string;
  bakedAt?: string;
  dispatchedAt?: string;
  deliveredAt?: string;
}

export interface CreateOrderRequest {
  storeId: string;
  customerName: string;
  customerPhone?: string;
  customerId?: string;
  items: Array<{
    menuItemId: string;
    name: string;
    quantity: number;
    price: number;
    variant?: string;
    customizations?: string[];
  }>;
  orderType: OrderType;
  paymentMethod?: PaymentMethod;
  deliveryAddress?: DeliveryAddress;
  specialInstructions?: string;
}

export interface UpdateOrderStatusRequest {
  orderId: string;
  status: OrderStatus;
}

export interface UpdateOrderItemsRequest {
  orderId: string;
  items: OrderItem[];
}

export interface AssignDriverRequest {
  orderId: string;
  driverId: string;
}

export interface UpdatePaymentRequest {
  orderId: string;
  status: PaymentStatus;
  transactionId?: string;
}

export interface UpdatePriorityRequest {
  orderId: string;
  priority: OrderPriority;
}

export interface CancelOrderRequest {
  orderId: string;
  reason?: string;
}

// UI Helper Types
export interface OrderFilters {
  storeId?: string;
  status?: OrderStatus;
  orderType?: OrderType;
  paymentStatus?: PaymentStatus;
  startDate?: string;
  endDate?: string;
  searchQuery?: string;
}

export interface OrderStats {
  totalOrders: number;
  activeOrders: number;
  completedOrders: number;
  cancelledOrders: number;
  totalRevenue: number;
  averageOrderValue: number;
}

// Status flow for kitchen workflow
// Note: After READY, the flow diverges based on order type:
// - DELIVERY: READY → DISPATCHED → DELIVERED
// - TAKEAWAY: READY → COMPLETED
// - DINE_IN: READY → SERVED
export const ORDER_STATUS_FLOW: OrderStatus[] = [
  'RECEIVED',
  'PREPARING',
  'OVEN',
  'BAKED',
  'READY',
  'DISPATCHED',  // DELIVERY only
  'DELIVERED',   // DELIVERY final state
  'COMPLETED',   // TAKEAWAY final state
  'SERVED'       // DINE_IN final state
];

// Helper to get next status in workflow
export const getNextOrderStatus = (currentStatus: OrderStatus): OrderStatus | null => {
  const currentIndex = ORDER_STATUS_FLOW.indexOf(currentStatus);
  if (currentIndex === -1 || currentIndex === ORDER_STATUS_FLOW.length - 1) {
    return null;
  }
  return ORDER_STATUS_FLOW[currentIndex + 1];
};

// Status display configurations
export const ORDER_STATUS_CONFIG: Record<OrderStatus, { label: string; color: string; icon: string }> = {
  RECEIVED: { label: 'Received', color: '#3b82f6', icon: '📋' },
  PREPARING: { label: 'Preparing', color: '#f59e0b', icon: '👨‍🍳' },
  OVEN: { label: 'In Oven', color: '#e53e3e', icon: '🔥' },
  BAKED: { label: 'Baked', color: '#f97316', icon: '🍕' },
  READY: { label: 'Ready', color: '#10b981', icon: '✅' },
  DISPATCHED: { label: 'Dispatched', color: '#8b5cf6', icon: '🚚' },
  DELIVERED: { label: 'Delivered', color: '#059669', icon: '📦' },
  COMPLETED: { label: 'Completed', color: '#059669', icon: '✅' },
  SERVED: { label: 'Served', color: '#10b981', icon: '🍽️' },
  CANCELLED: { label: 'Cancelled', color: '#6b7280', icon: '❌' },
};

// Order type display configurations
export const ORDER_TYPE_CONFIG: Record<OrderType, { label: string; color: string; icon: string }> = {
  DELIVERY: { label: 'Delivery', color: '#3b82f6', icon: '🚚' },
  TAKEAWAY: { label: 'Takeaway', color: '#10b981', icon: '🏪' },
  DINE_IN: { label: 'Dine-In', color: '#f59e0b', icon: '🍽️' },
};

// Payment status display configurations
export const PAYMENT_STATUS_CONFIG: Record<PaymentStatus, { label: string; color: string; icon: string }> = {
  PENDING: { label: 'Pending', color: '#f59e0b', icon: '⏳' },
  PAID: { label: 'Paid', color: '#10b981', icon: '✅' },
  FAILED: { label: 'Failed', color: '#ef4444', icon: '❌' },
  REFUNDED: { label: 'Refunded', color: '#6b7280', icon: '↩️' },
};

File: order-service/src/main/java/com/MaSoVa/order/entity/Order.java
package com.MaSoVa.order.entity;

import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.index.Indexed;

import java.time.LocalDateTime;
import java.util.List;

@Document(collection = "orders")
public class Order {

    @Id
    private String id;

    @Indexed(unique = true)
    private String orderNumber;

    private String customerId;
    private String customerName;
    private String customerPhone;

    @Indexed
    private String storeId;

    private List<OrderItem> items;

    private Double subtotal;
    private Double deliveryFee;
    private Double tax;
    private Double total;

    @Indexed
    private OrderStatus status;

    private OrderType orderType;

    private PaymentStatus paymentStatus;
    private PaymentMethod paymentMethod;
    private String paymentTransactionId;

    private Priority priority;

    private Integer preparationTime;
    private LocalDateTime estimatedDeliveryTime;

    private DeliveryAddress deliveryAddress;

    private String assignedDriverId;

    private String specialInstructions;

    @CreatedDate
    private LocalDateTime createdAt;

    @LastModifiedDate
    private LocalDateTime updatedAt;

    private LocalDateTime completedAt;

    private LocalDateTime cancelledAt;
    private String cancellationReason;

    // Kitchen workflow timestamps
    private LocalDateTime receivedAt;
    private LocalDateTime preparingStartedAt;
    private LocalDateTime ovenStartedAt;
    private LocalDateTime bakedAt;
    private LocalDateTime dispatchedAt;
    private LocalDateTime deliveredAt;

    // Quality checkpoints
    private List<QualityCheckpoint> qualityCheckpoints;

    // Actual preparation time tracking (in minutes)
    private Integer actualPreparationTime;
    private Integer actualOvenTime;

    // Make-table workflow
    private String assignedMakeTableStation;
    private String assignedKitchenStaffId;
    private String assignedKitchenStaffName;
    private LocalDateTime assignedToKitchenAt;

    // Constructors
    public Order() {}

    // Getters and Setters
    public String getId() { return id; }
    public void setId(String id) { this.id = id; }

    public String getOrderNumber() { return orderNumber; }
    public void setOrderNumber(String orderNumber) { this.orderNumber = orderNumber; }

    public String getCustomerId() { return customerId; }
    public void setCustomerId(String customerId) { this.customerId = customerId; }

    public String getCustomerName() { return customerName; }
    public void setCustomerName(String customerName) { this.customerName = customerName; }

    public String getCustomerPhone() { return customerPhone; }
    public void setCustomerPhone(String customerPhone) { this.customerPhone = customerPhone; }

    public String getStoreId() { return storeId; }
    public void setStoreId(String storeId) { this.storeId = storeId; }

    public List<OrderItem> getItems() { return items; }
    public void setItems(List<OrderItem> items) { this.items = items; }

    public Double getSubtotal() { return subtotal; }
    public void setSubtotal(Double subtotal) { this.subtotal = subtotal; }

    public Double getDeliveryFee() { return deliveryFee; }
    public void setDeliveryFee(Double deliveryFee) { this.deliveryFee = deliveryFee; }

    public Double getTax() { return tax; }
    public void setTax(Double tax) { this.tax = tax; }

    public Double getTotal() { return total; }
    public void setTotal(Double total) { this.total = total; }

    public OrderStatus getStatus() { return status; }
    public void setStatus(OrderStatus status) { this.status = status; }

    public OrderType getOrderType() { return orderType; }
    public void setOrderType(OrderType orderType) { this.orderType = orderType; }

    public PaymentStatus getPaymentStatus() { return paymentStatus; }
    public void setPaymentStatus(PaymentStatus paymentStatus) { this.paymentStatus = paymentStatus; }

    public PaymentMethod getPaymentMethod() { return paymentMethod; }
    public void setPaymentMethod(PaymentMethod paymentMethod) { this.paymentMethod = paymentMethod; }

    public String getPaymentTransactionId() { return paymentTransactionId; }
    public void setPaymentTransactionId(String paymentTransactionId) { this.paymentTransactionId = paymentTransactionId; }

    public Priority getPriority() { return priority; }
    public void setPriority(Priority priority) { this.priority = priority; }

    public Integer getPreparationTime() { return preparationTime; }
    public void setPreparationTime(Integer preparationTime) { this.preparationTime = preparationTime; }

    public LocalDateTime getEstimatedDeliveryTime() { return estimatedDeliveryTime; }
    public void setEstimatedDeliveryTime(LocalDateTime estimatedDeliveryTime) { this.estimatedDeliveryTime = estimatedDeliveryTime; }

    public DeliveryAddress getDeliveryAddress() { return deliveryAddress; }
    public void setDeliveryAddress(DeliveryAddress deliveryAddress) { this.deliveryAddress = deliveryAddress; }

    public String getAssignedDriverId() { return assignedDriverId; }
    public void setAssignedDriverId(String assignedDriverId) { this.assignedDriverId = assignedDriverId; }

    public String getSpecialInstructions() { return specialInstructions; }
    public void setSpecialInstructions(String specialInstructions) { this.specialInstructions = specialInstructions; }

    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }

    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }

    public LocalDateTime getCompletedAt() { return completedAt; }
    public void setCompletedAt(LocalDateTime completedAt) { this.completedAt = completedAt; }

    public LocalDateTime getCancelledAt() { return cancelledAt; }
    public void setCancelledAt(LocalDateTime cancelledAt) { this.cancelledAt = cancelledAt; }

    public String getCancellationReason() { return cancellationReason; }
    public void setCancellationReason(String cancellationReason) { this.cancellationReason = cancellationReason; }

    public LocalDateTime getReceivedAt() { return receivedAt; }
    public void setReceivedAt(LocalDateTime receivedAt) { this.receivedAt = receivedAt; }

    public LocalDateTime getPreparingStartedAt() { return preparingStartedAt; }
    public void setPreparingStartedAt(LocalDateTime preparingStartedAt) { this.preparingStartedAt = preparingStartedAt; }

    public LocalDateTime getOvenStartedAt() { return ovenStartedAt; }
    public void setOvenStartedAt(LocalDateTime ovenStartedAt) { this.ovenStartedAt = ovenStartedAt; }

    public LocalDateTime getBakedAt() { return bakedAt; }
    public void setBakedAt(LocalDateTime bakedAt) { this.bakedAt = bakedAt; }

    public LocalDateTime getDispatchedAt() { return dispatchedAt; }
    public void setDispatchedAt(LocalDateTime dispatchedAt) { this.dispatchedAt = dispatchedAt; }

    public LocalDateTime getDeliveredAt() { return deliveredAt; }
    public void setDeliveredAt(LocalDateTime deliveredAt) { this.deliveredAt = deliveredAt; }

    public List<QualityCheckpoint> getQualityCheckpoints() { return qualityCheckpoints; }
    public void setQualityCheckpoints(List<QualityCheckpoint> qualityCheckpoints) { this.qualityCheckpoints = qualityCheckpoints; }

    public Integer getActualPreparationTime() { return actualPreparationTime; }
    public void setActualPreparationTime(Integer actualPreparationTime) { this.actualPreparationTime = actualPreparationTime; }

    public Integer getActualOvenTime() { return actualOvenTime; }
    public void setActualOvenTime(Integer actualOvenTime) { this.actualOvenTime = actualOvenTime; }

    public String getAssignedMakeTableStation() { return assignedMakeTableStation; }
    public void setAssignedMakeTableStation(String assignedMakeTableStation) { this.assignedMakeTableStation = assignedMakeTableStation; }

    public String getAssignedKitchenStaffId() { return assignedKitchenStaffId; }
    public void setAssignedKitchenStaffId(String assignedKitchenStaffId) { this.assignedKitchenStaffId = assignedKitchenStaffId; }

    public String getAssignedKitchenStaffName() { return assignedKitchenStaffName; }
    public void setAssignedKitchenStaffName(String assignedKitchenStaffName) { this.assignedKitchenStaffName = assignedKitchenStaffName; }

    public LocalDateTime getAssignedToKitchenAt() { return assignedToKitchenAt; }
    public void setAssignedToKitchenAt(LocalDateTime assignedToKitchenAt) { this.assignedToKitchenAt = assignedToKitchenAt; }

    // Builder pattern
    public static Builder builder() { return new Builder(); }

    public static class Builder {
        private final Order order = new Order();

        public Builder id(String id) { order.id = id; return this; }
        public Builder orderNumber(String orderNumber) { order.orderNumber = orderNumber; return this; }
        public Builder customerId(String customerId) { order.customerId = customerId; return this; }
        public Builder customerName(String customerName) { order.customerName = customerName; return this; }
        public Builder customerPhone(String customerPhone) { order.customerPhone = customerPhone; return this; }
        public Builder storeId(String storeId) { order.storeId = storeId; return this; }
        public Builder items(List<OrderItem> items) { order.items = items; return this; }
        public Builder subtotal(Double subtotal) { order.subtotal = subtotal; return this; }
        public Builder deliveryFee(Double deliveryFee) { order.deliveryFee = deliveryFee; return this; }
        public Builder tax(Double tax) { order.tax = tax; return this; }
        public Builder total(Double total) { order.total = total; return this; }
        public Builder status(OrderStatus status) { order.status = status; return this; }
        public Builder orderType(OrderType orderType) { order.orderType = orderType; return this; }
        public Builder paymentStatus(PaymentStatus paymentStatus) { order.paymentStatus = paymentStatus; return this; }
        public Builder paymentMethod(PaymentMethod paymentMethod) { order.paymentMethod = paymentMethod; return this; }
        public Builder priority(Priority priority) { order.priority = priority; return this; }
        public Builder preparationTime(Integer preparationTime) { order.preparationTime = preparationTime; return this; }
        public Builder estimatedDeliveryTime(LocalDateTime estimatedDeliveryTime) { order.estimatedDeliveryTime = estimatedDeliveryTime; return this; }
        public Builder deliveryAddress(DeliveryAddress deliveryAddress) { order.deliveryAddress = deliveryAddress; return this; }
        public Builder specialInstructions(String specialInstructions) { order.specialInstructions = specialInstructions; return this; }
        public Builder receivedAt(LocalDateTime receivedAt) { order.receivedAt = receivedAt; return this; }
        public Builder qualityCheckpoints(List<QualityCheckpoint> qualityCheckpoints) { order.qualityCheckpoints = qualityCheckpoints; return this; }

        public Order build() { return order; }
    }

    // Enums
    public enum OrderStatus {
        RECEIVED,
        PREPARING,
        OVEN,
        BAKED,
        DISPATCHED,
        DELIVERED,
        CANCELLED
    }

    public enum OrderType {
        DINE_IN,
        TAKEAWAY,
        DELIVERY
    }

    public enum PaymentStatus {
        PENDING,
        PAID,
        FAILED,
        REFUNDED
    }

    public enum PaymentMethod {
        CASH,
        CARD,
        UPI,
        WALLET
    }

    public enum Priority {
        NORMAL,
        URGENT
    }
}
